src/eos-shard-writer-v2.c

```diff
@@ -262,45 +262,80 @@ eos_shard_writer_v2_new_for_fd (int fd)
   return EOS_SHARD_WRITER_V2 (g_object_new (EOS_SHARD_TYPE_WRITER_V2, "fd", fd, NULL));
 }
 
+static int
+compress_blob_to_tmp (GInputStream *file_stream)
+{
+  char tmpl[] = "/tmp/shardXXXXXX";
+  int fd = mkstemp (tmpl);
+  unlink (tmpl);
+
+  /* Compress the given GInputStream to the tmpfile. */
+
+  g_autoptr(GZlibCompressor) compressor = g_zlib_compressor_new (G_ZLIB_COMPRESSOR_FORMAT_ZLIB, -1);
+  g_autoptr(GInputStream) stream = g_converter_input_stream_new (G_INPUT_STREAM (file_stream), G_CONVERTER (compressor));
+
+  uint8_t buf[4096*4];
+  int size;
+  while ((size = g_input_stream_read (stream, buf, sizeof (buf), NULL, NULL)) != 0) {
+    g_assert (write (fd, buf, size) >= 0);
+  }
+
+  return fd;
+}
+
 static void
-write_blob (int fd, struct eos_shard_writer_v2_blob_entry *blob, GFile *file)
+write_blob (EosShardWriterV2 *self, struct eos_shard_writer_v2_blob_entry *blob, GFile *file)
 {
   g_autoptr(GError) error = NULL;
 
-  GFileInputStream *file_stream = g_file_read (file, NULL, &error);
+  g_autoptr(GFileInputStream) file_stream = g_file_read (file, NULL, &error);
   if (!file_stream) {
     g_error ("Could not read from %s: %s", g_file_get_path (file), error->message);
     return;
   }
 
-  g_autoptr(GInputStream) stream;
-
+  int blob_fd;
   if (blob->sblob.flags & EOS_SHARD_BLOB_FLAG_COMPRESSED_ZLIB) {
-    g_autoptr(GZlibCompressor) compressor = g_zlib_compressor_new (G_ZLIB_COMPRESSOR_FORMAT_ZLIB, -1);
-    stream = g_converter_input_stream_new (G_INPUT_STREAM (file_stream), G_CONVERTER (compressor));
-    g_object_unref (file_stream);
+    blob_fd = compress_blob_to_tmp (G_INPUT_STREAM (file_stream));
   } else {
-    stream = G_INPUT_STREAM (file_stream);
+    blob_fd = g_file_descriptor_based_get_fd (G_FILE_DESCRIPTOR_BASED (file_stream));
   }
 
+  /* Find the size by seeking... */
+  uint64_t blob_size = lseek (blob_fd, 0, SEEK_END);
```

Comment:
Ignore this one - stat looks to be about the same in complexity to use and gives the same results